File: trees/go/solution02.go
package leetcode

func isValidBST(root *TreeNode) bool {
	return helper(root, nil, nil)
}

func helper(node *TreeNode, min *int, max *int) bool {
	if node == nil {
		return true
	}

	if (min != nil && node.Val <= *min) || (max != nil && node.Val >= *max) {
		return false
	}

	return helper(node.Left, min, &node.Val) && helper(node.Right, &node.Val, max)
}

File: trees/java/Solution02.java

public class Solution02 {

    public class TreeNode {

        TreeNode left;
        TreeNode right;
        int val;
    }

    public boolean isValidBST(TreeNode root) {
        return helper(root, null, null);
    }

    public boolean helper(TreeNode node, Integer min, Integer max) {
        if (node == null) {
            return true;
        }

        if ((min != null && node.val <= min) || (max != null && node.val >= max)) {
            return false;
        }

        return helper(node.left, min, node.val) && helper(node.right, node.val, max);
    }
}
